packages/flutter/lib/src/cupertino/expansion_tile.dart

```diff
@@ -0,0 +1,230 @@
+  ///
+  /// When the [CupertinoExpansionTile] expands, the child appears fully extended
+  /// and fades into view. When the [CupertinoExpansionTile] collapses, the child
+  /// remains fully extended and fades out of view.
+  fade,
+
+  /// Transition by scrolling [CupertinoExpansionTile.child] under the header.
+  ///
+  /// When the [CupertinoExpansionTile] expands, the child scrolls from under the
+  /// header until it becomes fully extended. When the [CupertinoExpansionTile]
+  /// collapses, the child scrolls under the header until it is fully collapsed.
+  scroll,
+}
+
+/// A single-line [CupertinoListTile] with an expansion arrow icon that expands
+/// or collapses the tile to reveal or hide the [child].
+///
+/// See also:
+///
+///  * [ExpansionTile], the Material Design equivalent.
+///  * [CupertinoListSection], useful for creating an expansion tile [child].
+///  * [CupertinoListTile], the header of a [CupertinoExpansionTile].
+class CupertinoExpansionTile extends StatefulWidget {
+  /// Creates a single-line [CupertinoListTile] with an expansion arrow icon
+  /// that expands or collapses the tile to reveal or hide the [child].
+  const CupertinoExpansionTile({
+    super.key,
+    required this.title,
+    required this.child,
+    this.controller,
+    this.transitionMode = ExpansionTileTransitionMode.fade,
+  });
+
+  /// A [title] is used to convey the central information. Usually a [Text].
+  final Widget title;
+
+  /// Programmatically expands and collapses the [CupertinoExpansionTile].
+  ///
+  /// In cases where control over the tile's state is needed from a
+  /// callback triggered by a widget within the tile, [ExpansibleController.of]
+  /// may be more convenient than supplying a controller.
+  final ExpansibleController? controller;
+
+  /// The body of the [CupertinoExpansionTile].
+  final Widget child;
+
+  /// How the [CupertinoExpansionTile] should transition its child between its
+  /// collapsed state and its expanded state.
+  ///
+  /// Defaults to [ExpansionTileTransitionMode.fade].
+  final ExpansionTileTransitionMode transitionMode;
+
+  @override
+  State<CupertinoExpansionTile> createState() => _CupertinoExpansionTileState();
+}
+
+class _CupertinoExpansionTileState extends State<CupertinoExpansionTile> {
+  final GlobalKey _headerKey = GlobalKey();
+  final OverlayPortalController _fadeController = OverlayPortalController();
+  static final Animatable<double> _quarterTween = Tween<double>(begin: 0.0, end: 0.25);
+
+  late ExpansibleController _tileController;
+  late Animation<double> _iconTurns;
+  late Offset _headerOffset;
+
+  @override
+  void initState() {
+    super.initState();
+    _tileController = widget.controller ?? ExpansibleController();
+  }
+
+  @override
+  void dispose() {
+    if (widget.controller == null) {
+      _tileController.dispose();
+    }
+    super.dispose();
+  }
+
+  Widget? _buildIcon(BuildContext context, Animation<double> animation) {
```

Comment:
Nit: Would it be any cleaner if you break out some of these builder methods out into separate widget(s)? I'm not sure in this case so it's up to you.

Reply:
Yes that's what I was thinking. Leaving as-is sounds good to me.